wallet/src/signer/ledger_signer/mod.rs

```diff
@@ -0,0 +1,1094 @@
+    #[error("Derivation path is too long to send to Ledger")]
+    PathToLong,
+    #[error("Invalid public key returned from Ledger")]
+    InvalidKey,
+    #[error("The file being loaded is a software wallet and does not correspond to the connected hardware wallet")]
+    WalletFileIsSoftwareWallet,
+    #[error("Public keys mismatch - wrong device or passphrase")]
+    HardwareWalletDifferentMnemonicOrPassphrase,
+    #[error("A multisig signature was returned for a single address from Device")]
+    MultisigSignatureReturned,
+    #[error("Multiple signatures returned for a single address from Device")]
+    MultipleSignaturesReturned,
+    #[error("Missing multisig index for signature returned from Device")]
+    MissingMultisigIndexForSignature,
+    #[error("Signature error: {0}")]
+    SignatureError(#[from] SignatureError),
+}
+
+struct StandaloneInput {
+    multisig_idx: Option<u32>,
+    private_key: PrivateKey,
+}
+
+type StandaloneInputs = BTreeMap</*input index*/ u32, Vec<StandaloneInput>>;
+
+#[async_trait]
+pub trait LedgerFinder {
+    type Ledger;
+
+    async fn find_ledger_device_from_db<T: WalletStorageReadLocked + Send>(
+        &self,
+        db_tx: &mut T,
+    ) -> SignerResult<(Self::Ledger, LedgerData)>;
+}
+
+pub struct LedgerSigner<L, P> {
+    chain_config: Arc<ChainConfig>,
+    client: Arc<Mutex<L>>,
+    sig_aux_data_provider: std::sync::Mutex<Box<dyn SigAuxDataProvider + Send>>,
+    provider: P,
+}
+
+impl<L, P> LedgerSigner<L, P>
+where
+    L: Exchange + Send,
+    P: LedgerFinder<Ledger = L>,
+{
+    pub fn new(chain_config: Arc<ChainConfig>, client: Arc<Mutex<L>>, provider: P) -> Self {
+        Self::new_with_sig_aux_data_provider(
+            chain_config,
+            client,
+            Box::new(make_true_rng()),
+            provider,
+        )
+    }
+
+    pub fn new_with_sig_aux_data_provider(
+        chain_config: Arc<ChainConfig>,
+        client: Arc<Mutex<L>>,
+        sig_aux_data_provider: Box<dyn SigAuxDataProvider + Send>,
+        provider: P,
+    ) -> Self {
+        Self {
+            chain_config,
+            client,
+            sig_aux_data_provider: std::sync::Mutex::new(sig_aux_data_provider),
+            provider,
+        }
+    }
+
+    /// Calls initialize on the device with the current session_id.
+    ///
+    /// If the operation fails due to an USB error (which may indicate a lost connection to the device),
+    /// the function will attempt to reconnect to the Ledger device once before returning an error.
+    async fn check_session<T: WalletStorageReadLocked + Send>(
+        &mut self,
+        db_tx: &mut T,
+        key_chain: &impl AccountKeyChains,
+    ) -> SignerResult<()> {
+        let mut client = self.client.lock().await;
+
+        loop {
+            match get_app_name(&mut *client).await {
+                Ok(_) => return Ok(()),
+                Err(ledger_lib::Error::Timeout) => {
+                    continue;
+                }
```

Comment:
I don't think it's a good idea to ignore timeouts indefinitely. Perhaps we should only allow a few timeouts and then fail?
But why do we even need to handle a timeout differently that other connection errors that a re handled below?